File: CpuA32/symbolic.py
"""
Convert ARM32 instruction into human readable form suitable to
be processed by an assembler.
"""

from typing import List, Tuple, Set
import struct
import CpuA32.opcode_tab as a32
from Elf import enum_tab


def _EmitReloc(ins: a32.Ins, pos: int) -> str:
    if ins.reloc_kind == enum_tab.RELOC_TYPE_ARM.JUMP24:
        assert ins.is_local_sym, f"expected local symbol"
        return f"expr:jump24:{ins.reloc_symbol}"
    elif ins.reloc_kind == enum_tab.RELOC_TYPE_ARM.MOVT_ABS:
        loc = "loc_" if ins.is_local_sym else ""
        offset = "" if ins.operands[pos] == 0 else f":{ins.operands[pos]}"
        return f"expr:{loc}movt_abs:{ins.reloc_symbol}{offset}"
    elif ins.reloc_kind == enum_tab.RELOC_TYPE_ARM.MOVW_ABS_NC:
        loc = "loc_" if ins.is_local_sym else ""
        offset = "" if ins.operands[pos] == 0 else f":{ins.operands[pos]}"
        return f"expr:{loc}movw_abs_nc:{ins.reloc_symbol}{offset}"
    elif ins.reloc_kind == enum_tab.RELOC_TYPE_ARM.CALL:
        return f"expr:call:{ins.reloc_symbol}"
    else:
        assert False


def _FloatFrom32BitRepresentation(data: int) -> float:
    return struct.unpack('<f', int.to_bytes(data, 4, "little"))[0]


def _SymbolizeOperand(ok: a32.OK, data: int) -> str:
    """Convert an operand in integer form as found in  arm.Ins to a string

    (This does not handle relocation expressions.)
    """
    data = a32.DecodeOperand(ok, data)
    assert data >= 0, f"expected unsigned for {ok}: {data}"
    t = a32.FIELD_DETAILS[ok]
    if t.kind == a32.FK.LIST:
        return t.names[data]
    elif t.kind == a32.FK.FLT_CUSTOM:
        # we only care about the float aspect
        data = _FloatFrom32BitRepresentation(data)
        return str(data)
    elif t.kind == a32.FK.INT_SIGNED_CUSTOM or t.kind == a32.FK.INT_SIGNED:
        # we only care about the signed aspect
        if data >= 1 << 31:
            data -= (1 << 32)
        return str(data)
    elif t.kind == a32.FK.INT_HEX:
        return t.prefix + hex(data)
    else:
        return t.prefix + str(data)


_REG_REWRITES = {
    "r10": "sl",
    "r11": "fp",
    "r12": "ip",
    "r14": "lr",
}


def _FloatTo32BitRepresentation(num: float) -> int:
    b = struct.pack('<f', num)
    assert len(b) == 4
    return int.from_bytes(b, "little")


def _UnsymbolizeOperand(ok: a32.OK, op: str) -> int:
    """Convert a symbolized operand  into an int suitable for arm.Ins

    This does not handle relocation expressions.
    """
    t = a32.FIELD_DETAILS[ok]
    if t.prefix and op.startswith(t.prefix):
        op = op[len(t.prefix):]
    if t.kind == a32.FK.LIST:
        op = _REG_REWRITES.get(op, op)
        data = t.names.index(op)
    elif t.kind == a32.FK.FLT_CUSTOM:
        # we only care about the float aspect
        data = _FloatTo32BitRepresentation(float(op))
    elif t.kind == a32.FK.INT_SIGNED_CUSTOM or t.kind == a32.FK.INT_SIGNED:
        # we only care about the signed aspect
        data = int(op, 0)
        if data < 0:
            data += 1 << 32
    else:
        data = int(op, 0)
    assert data is not None and data >= 0, f"bad data {ok}: {op}"
    return a32.EncodeOperand(ok, data)


_RELOC_KIND_MAP = {
    "jump24": (enum_tab.RELOC_TYPE_ARM.JUMP24, True), # always local
    #
    "abs32": (enum_tab.RELOC_TYPE_ARM.ABS32, False),
    "call": (enum_tab.RELOC_TYPE_ARM.CALL, False),
    "movw_abs_nc": (enum_tab.RELOC_TYPE_ARM.MOVW_ABS_NC, False),
    "movt_abs": (enum_tab.RELOC_TYPE_ARM.MOVT_ABS, False),
    #
    "loc_abs32": (enum_tab.RELOC_TYPE_ARM.ABS32, True),
    "loc_call": (enum_tab.RELOC_TYPE_ARM.CALL, True),
    "loc_movw_abs_nc": (enum_tab.RELOC_TYPE_ARM.MOVW_ABS_NC, True),
    "loc_movt_abs": (enum_tab.RELOC_TYPE_ARM.MOVT_ABS, True),
}

_RELOC_OK: Set[a32.OK] = {a32.OK.SIMM_0_23, a32.OK.IMM_0_11_16_19}


def InsSymbolize(ins: a32.Ins) -> Tuple[str, List[str]]:
    """Convert all the operands in an arm.Ins to strings including relocs
    """
    ops = []
    for pos, (ok, value) in enumerate(zip(ins.opcode.fields, ins.operands)):
        if ok in _RELOC_OK and ins.has_reloc() and ins.reloc_pos == pos:
            # Note: we essentially store the "addend" here
            ops.append(_EmitReloc(ins, pos))
        else:
            ops.append(_SymbolizeOperand(ok, value))

    return ins.opcode.name, ops


def InsFromSymbolized(mnemonic, token: List[str]) -> a32.Ins:
    """Takes textual form of an A32 instruction (in systematic notation) and parses it into a Ins

     * Supports relocatable expressions
     * Adds missing "al" predicate if necessary
     Example input:
     "add_regimm", ["r4", "r4", "r0", "lsl", "0"]
     """
    opcode: a32.Opcode = a32.Opcode.name_to_opcode[mnemonic]
    # CodeGenA32 relies on this
    if opcode.HasPred() and len(token) == len(opcode.fields) - 1:
        token = ["al"] + token
    ins = a32.Ins(opcode)
    assert len(token) == len(opcode.fields), f"expected {len(opcode.fields)} operands in {mnemonic} {token}"
    for pos, (t, ok) in enumerate(zip(token, opcode.fields)):
        if t.startswith("expr:"):
            # expr strings have the form expr:<rel-kind>:<symbol>:<addend>, e.g.:
            #   expr:movw_abs_nc:string_pointers:5
            #   expr:call:putchar
            rel_token = t.split(":")

            ins.set_reloc(*_RELOC_KIND_MAP[rel_token[1]], pos, rel_token[2])
            addend = 0 if len(rel_token) ==3 else  int(rel_token[3], 0)
            ins.operands.append(addend)
        else:
            ins.operands.append(_UnsymbolizeOperand(ok, t))
    return ins

File: CpuA32/symbolic.cc
// (c) Robert Muth - see LICENSE for more info

#include "CpuA32/symbolic.h"
#include "Util/assert.h"
#include "Util/parse.h"

#include <cstring>
#include <string_view>

namespace cwerg::a32 {
namespace {

char* strappenddec(char* dst, uint32_t n) {
  if (n <= 0x7fffffff) {
    return dst + ToDecString(n, dst).size();
  } else {
    *dst++ = '-';
    return dst + ToDecString(~n + 1, dst).size();
  }
}

char* strappendhex(char* dst, uint32_t n) {
  return dst + ToHexString(n, dst).size();
}

char* strappend(char* dst, std::string_view src) {
  std::memcpy(dst, src.data(), src.size());
  dst[src.size()] = 0;
  return dst + src.size();
}

char* SymbolizeOperand(char* buffer, uint32_t data, OK ok) {
  const FieldInfo& fi = FieldInfoTable[uint8_t(ok)];
  const uint32_t x = a32::DecodeOperand(data, ok);
  switch (fi.kind) {
    default:
    case FK::NONE:
      ASSERT(false, "unreachable");
      return nullptr;
    case FK::INT_SIGNED_CUSTOM:
      return strappenddec(buffer, x);
    case FK::FLT_CUSTOM:
      ASSERT(x == 0, "");
      return strappend(buffer, "0.0");
    case FK::LIST:
      ASSERT(x < fi.num_names, "out of range for list " << x << " " << (int)ok);
      return strappend(buffer, fi.names[x]);
    case FK::INT_SIGNED:
      return strappenddec(buffer, x);
    case FK::INT:
      buffer = strappend(buffer, fi.prefix);
      return strappenddec(buffer, x);
    case FK::INT_HEX:
      buffer = strappend(buffer, fi.prefix);
      buffer = strappend(buffer, "0x");
      return strappendhex(buffer, x);
  }
}

bool has_prefix(std::string_view name, std::string_view prefix) {
  return name.substr(0, prefix.size()) == prefix;
}

uint32_t UnsymbolizeOperandString(const FieldInfo& fi, std::string_view s) {
  switch (fi.kind) {
    default:
    case FK::NONE:
      ASSERT(false, "unreachable");
      return kEncodeFailure;
    case FK::INT_SIGNED:
    case FK::INT_SIGNED_CUSTOM: {
      auto val = ParseInt<int32_t>(s);
      if (!val) return kEncodeFailure;
      return val.value();
    }
    case FK::FLT_CUSTOM:
      if (s != "0.0" && s != ".0" && s != "0") return kEncodeFailure;
      return 0;  // this is really the bit representation of float(0.0)
    case FK::LIST:
      for (unsigned i = 0; i < fi.num_names; ++i) {
        if (s == fi.names[i]) return i;
      }
      return kEncodeFailure;
    case FK::INT:
    case FK::INT_HEX: {
      auto val = ParseInt<uint32_t>(s);
      if (!val) return kEncodeFailure;
      return val.value();
    }
  }
}

uint32_t UnsymbolizeOperand(OK ok, std::string_view s) {
  const FieldInfo& fi = FieldInfoTable[uint8_t(ok)];
  if (!fi.prefix.empty() && has_prefix(s, fi.prefix)) {
    s.remove_prefix(fi.prefix.size());
  } else if (s == "r10")
    s = "sl";
  else if (s == "r11")
    s = "fp";
  else if (s == "r12")
    s = "ip";

  return a32::EncodeOperand(UnsymbolizeOperandString(fi, s), ok);
}

const constexpr struct {
  std::string_view expr_kind;
  elf::RELOC_TYPE_ARM reloc_kind;
  bool local;
} kExprToReloc[] = {
    // order by frequency
    {"jump24", elf::RELOC_TYPE_ARM::JUMP24, true},
    {"call", elf::RELOC_TYPE_ARM::CALL, false},
    {"abs32", elf::RELOC_TYPE_ARM::ABS32, false},
    {"movw_abs_nc", elf::RELOC_TYPE_ARM::MOVW_ABS_NC, false},
    {"movt_abs", elf::RELOC_TYPE_ARM::MOVT_ABS, false},
    {"loc_movw_abs_nc", elf::RELOC_TYPE_ARM::MOVW_ABS_NC, true},
    {"loc_movt_abs", elf::RELOC_TYPE_ARM::MOVT_ABS, true},
};

bool HandleRelocation(std::string_view expr, unsigned pos, Ins* ins) {
  auto expr_op = ParseExpressionOp(expr);
  if (!expr_op) return false;
  ins->operands[pos] = expr_op->offset;
  for (const auto& x : kExprToReloc) {
    if (x.expr_kind == expr_op->reloc_name) {
      ins->set_reloc(x.reloc_kind, x.local, pos, expr_op->symbol_name);
      return true;
    }
  }
  return false;
}

void SymbolizeReloc(char* cp, const Ins& ins, uint32_t addend) {
  cp = strappend(cp, "expr:");
  switch (ins.reloc_kind) {
    case elf::RELOC_TYPE_ARM::JUMP24:
      ASSERT(ins.is_local_sym, "");
      cp = strappend(cp, "jump24:");
      cp = strappend(cp, ins.reloc_symbol);
      break;
    case elf::RELOC_TYPE_ARM::CALL:
      cp = strappend(cp, "call:");
      cp = strappend(cp, ins.reloc_symbol);
      break;
    case elf::RELOC_TYPE_ARM::MOVT_ABS:
      cp = strappend(cp, ins.is_local_sym ? "loc_movt_abs:" : "movt_abs:");
      cp = strappend(cp, ins.reloc_symbol);
      break;
    case elf::RELOC_TYPE_ARM::MOVW_ABS_NC:
      cp =
          strappend(cp, ins.is_local_sym ? "loc_movw_abs_nc:" : "movw_abs_nc:");
      cp = strappend(cp, ins.reloc_symbol);

      break;
    default:
      ASSERT(false, "");
  }
  if (addend != 0) {
    *cp++ = ':';
    cp = strappenddec(cp, addend);
  }
}

}  // namespace

std::string_view InsSymbolize(const Ins& ins, std::vector<std::string>* ops) {
  char buffer[128];
  for (unsigned i = 0; i < ins.opcode->num_fields; ++i) {
    if (ins.has_reloc() && i == ins.reloc_pos) {
      SymbolizeReloc(buffer, ins, ins.operands[i]);
    } else {
      SymbolizeOperand(buffer, ins.operands[i], ins.opcode->fields[i]);
    }
    ops->emplace_back(buffer);
  }
  return ins.opcode->enum_name;
}

bool InsFromSymbolized(const std::vector<std::string_view>& token, Ins* ins) {
  ins->opcode = FindOpcodeForMnemonic(token[0]);
  if (ins->opcode == nullptr) {
    std::cerr << "unknown opcode " << token[0] << "\n";
    return false;
  }
  uint32_t operand_count = 0;
  // CodeGenA32 relies on this
  if (token.size() == ins->opcode->num_fields) {
    ins->operands[operand_count++] = 14;  // predicate 'al'
  }
  ASSERT(token.size() - 1 + operand_count == ins->opcode->num_fields, "");
  for (unsigned i = 1; i < token.size(); ++i, ++operand_count) {
    if (token[i].substr(0, 5) == "expr:") {
      if (!HandleRelocation(token[i].substr(5), operand_count, ins)) {
        std::cerr << "malformed relocation expression " << token[i] << "\n";
        return false;
      }
    } else {
      const OK ok = ins->opcode->fields[operand_count];
      auto val = UnsymbolizeOperand(ok, token[i]);
      if (val == kEncodeFailure) {
        std::cerr << "cannot parse " << token[i] << " "
                  << "\n";
        return false;
      }
      ins->operands[operand_count] = val;
    }
  }
  return true;
}

std::string SymbolizeRegListMask(uint32_t mask) {
  std::string regs = "{";
  std::string_view sep;
  for (unsigned int i = 0; i < 16; ++i) {
    if (mask & (1U << i)) {
      regs += sep;
      regs += EnumToString(a32::REG(i));
      sep = ", ";
    }
  }
  regs += "}";
  return regs;
}

std::string SymbolizeRegRange(uint32_t start_reg_no,
                              uint32_t count,
                              bool single) {
  ASSERT(count > 0, "");
  std::string_view start_reg = single ? EnumToString(a32::SREG(start_reg_no))
                                      : EnumToString(a32::DREG(start_reg_no));
  std::string out = "{";
  out += start_reg;
  if (count > 1) {
    uint32_t end_reg_no = start_reg_no + count - 1;
    std::string_view end_reg = single ? EnumToString(a32::SREG(end_reg_no))
                                      : EnumToString(a32::DREG(end_reg_no));
    out += "-";
    out += end_reg;
  }
  return out + "}";
}

}  // namespace cwerg::a32
